Calling conventions for the lazily bound functions. (#190)
riscv-non-isa/riscv-elf-psabi-doc: riscv-elf.md

to address literals in the GOT.  This code model is position independent.
Does not apply to the ILP32 ABIs.

## <a name=dynamic-linking /> Dynamic linking

Lazy binding must follow the standard calling conventions. The resolver in the

> dynamic linker will save/restore `a0-a7` for integer calling convention and

Reviewer: I'd cut out the bit about save/restore, and just say that binding must follow the standard calling conventions.  IMO it's really up to the C library to determine what it needs to save/restore, and while we do currently just save/restore both X and F args in glibc that's just because we haven't figured out a reliable way to exclude the F extensions from the dynamic loader.  These are all build system concerns in glibc and other C libraries may not have those concerns so I don't want to force them to spin up the FPU if they don't need to.

save/restore `fa0-fa7` for hardware floating-point calling convention to avoid
ruining the arguments before jumping to the resolved function. If there is any
need to support non-standard calling convention for lazy binding, the symbol

> needs to be decorated using `STO_RISCV_BIND_NOW`. With the attribute in the

Reviewer 2: `s/with the/with this/`

to address literals in the GOT.  This code model is position independent.
Does not apply to the ILP32 ABIs.

## <a name=dynamic-linking /> Dynamic linking

Lazily bound functions must follow the standard calling conventions. The
resolver in the dynamic linker will save/restore `a0-a7` for integer calling
convention and save/restore `fa0-fa7` for hardware floating-point calling
convention to avoid ruining the arguments before jumping to the resolved
function. If there is any need to support non-standard calling convention for

> lazily bound functions, the symbol needs to be decorated using

Reviewer 2: `s/decorated/annotated/`

to address literals in the GOT.  This code model is position independent.
Does not apply to the ILP32 ABIs.

## <a name=dynamic-linking></a>Dynamic Linking

Run-time linkers that use lazy binding must preserve all argument registers
used in the standard calling convention for the ABI in use. Any functions that

> use additional argument registers must be annoted with `STO_RISCV_VARIANT_CC`,

Reviewer 3: ```suggestion
use additional argument registers must be annotated with `STO_RISCV_VARIANT_CC`,
```

Reviewer 4: Thanks, late night typos strike again...

## <a name=symbol-table></a>Symbol Table

* st_other: The lower 2 bits are used to specify a symbol's visibility. The
  remaining 6 bits have no defined meaning in gABI. We use the highest bit for
  the eager binding semantic of the function call with the non-standard calling

>   convention or any other special purpose to avoid going through the resolver.

Reviewer 4: See below

  RISC-V specific `st_other` flags
  ----------------------------
  Name                 | Mask
  ---------------------|------
  STO_RISCV_VARIANT_CC | 0x80

  If `STO_RISCV_VARIANT_CC` is set, the dynamic linker will always resolve the
  symbol during program loading. The resolved symbol address will be filled
  into GOT entry regardless if `LD_BIND_NOW` is set or not under dynamic
  linking.

  Static linkers must set the flag for the symbol following the eager binding
  semantic in the dynamic symbol table and add a `DT_RISCV_VARIANT_CC` dynamic

>   tag in the Dynamic Section of the object.

Reviewer 4: This is overly prescriptive. If a dynamic linker wants to save and restore every single register, it can do so and then continue to lazily bind everything. It's likely a bad idea, but it's a correct implementation. This should only state what the requirements are; commentary about this likely implying eager binding belongs in a `> NOTE:`.

> NOTE:

>   Vector registers have variant size. It depends on the hardware implementation.

Reviewer 5: "variant size" doesn't read well to me. If it's a standard term then ignore this suggestion, but something like:

```suggestion
  Vector registers have a variable size depending on the hardware implementation.
```

>   To save/restore all these vector arguments in the resolver may occupy a large

Reviewer 5: This sentence is missing something. Either of these read better to me:
```suggestion
  To save/restore all these vector arguments, the resolver may occupy a large
```
or
```suggestion
  Saving/restoring all these vector arguments in the resolver may occupy a large
```

## <a name=symbol-table></a>Symbol Table

* st_other: The lower 2 bits are used to specify a symbol's visibility. The
  remaining 6 bits have no defined meaning in gABI. We use the highest bit for
  the eager binding semantic of the function call with the non-standard calling
  convention or any other special purpose to avoid going through the resolver.
  For example, vector registers have variant size. It may be from 128 bits to

>   4096 bits or larger. It depends on the hardware implementation. To

Reviewer: Much as above, I'd prefer to keep some of this extra wording out of the ABI.  We don't really have a way to split out the normative text from comments, and while having these examples is nice to explain why we're making certain decisions it could be construed as an actual ABI requirement.  From "128 bits to 4096 bits or larger" is pretty inclusive, but it's just unnecessarry.

Maybe the right answer here is to add some specific commentary decorations?  We've got a handful of these in the ABI doc right now so it'd be a nice cleanup.

Reviewer 4: My thinking for commentary while we aren't using asciidoc is something like:

> NOTE: Bla bla bla

## <a name=symbol-table></a>Symbol Table

* st_other: The lower 2 bits are used to specify a symbol's visibility. The
  remaining 6 bits have no defined meaning in gABI. We use the highest bit for
  the bind now semantic of the function call with the non-standard calling
  convention or any other special purpose to avoid going through the resolver.
  For example, vector registers have variant size. It may be from 128 bits to
  4096 bits or larger. It depends on the hardware implementations. To
  save/restore all these vector arguments in the resolver will occupy a large
  portion of stack space. In addition, the efficiency is also a factor to
  consider for vector calls.

  RISC-V specific `st_other` flags
  --------------------------
  Name               | Mask
  -------------------|------
  STO_RISCV_BIND_NOW | 0x80

  If `STO_RISCV_BIND_NOW` is set, the dynamic linker will resolve the symbol
  during program loading. The resolved symbol address will be filled into GOT
  entry regardless `LD_BIND_NOW` is set or not under dynamic linking for PIC

>   code.

Reviewer 2: `s/for PIC code//`

## <a name=symbol-table></a>Symbol Table

* st_other: The lower 2 bits are used to specify a symbol's visibility. The

>   remaining 6 bits have no defined meaning in the ELF gABI. We use the highest

Reviewer 6: The wording needs to be careful but the current one is fine.

https://groups.google.com/g/generic-abi/c/ytficX3WEWg/m/TfPyG4MFCQAJ

There are no RISC-V specific definitions relating to ELF note sections.

## <a name=dynamic-section></a>Dynamic Section

RISC-V specific dynamic array tags
---------------------------------------------------------------------------------
Name                | Value      | d_un  | Executable        | Shared Object
--------------------|------------|-------|-------------------|-------------------

> DT_RISCV_VARIANT_CC | 0x70000001 | d_val | Platform specific | Platform specific

Reviewer 7: Curious why skip `0x70000000 (DT_LOPROC + 0)`? I saw some targets like Alpha, PPC, PPC64, IA64 using that value.

Reviewer 7: Oh, after checking with AArch64's glibc implementation[1], I realize that's an implementation optimization, we can use `sym->st_other & DT_RISCV_VARIANT_CC` to check that.

[1] https://github.com/riscv/riscv-glibc/commit/82bc69c012838a381c4167c156a06f4598f34227#diff-5c636a3d53a784449673f0122b405bd56513de968c2991e6ba2fd894fb34e930R412

Author: No, DT_RISCV_VARIANT_CC is not used to check sym->st_other. AFAIK, there is an encoding rule for dynamic tags in gABI as follow.

"A tag whose value is an even number indicates a dynamic section entry that uses d_ptr. A tag whose value is an odd number indicates a dynamic section entry that uses d_val or that uses neither d_ptr nor d_val."

That's why it is encoded as 0x70000001.

There are no RISC-V specific definitions relating to ELF note sections.

## <a name=dynamic-section></a>Dynamic Section

The defined processor-specific dynamic array tags are listed in the following
table.

Name                | Value      | d_un  | Executable        | Shared Object
:------------------ | :--------- | :---- | :---------------- | :-----------------
DT_RISCV_VARIANT_CC | 0x70000001 | d_val | Platform specific | Platform specific

> An object with the dynamic tag `DT_RISCV_VARIANT_CC` has one or more

Reviewer 6: This sentence does not imply that `DT_RISCV_VARIANT_CC` must be present.

Reviewer 7: How about this?
```
An object must having  dynamic tag `DT_RISCV_VARIANT_CC` if having any `R_RISCV_JUMP_SLOT` relocations against symbols with the `STO_RISCV_VARIANT_CC` attribute.
```
